[PyTorch][Distributed] Enable Reduce Scatter, all_to_all_single and modify all_to_all for sharded linear with more test cases.

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68096
* __->__ #68786

To enable the auto grad for the sharded linear, we find we need to make some changes to the current nn function api (c10d api with auto grad enabled). So we made the following several changes:

1. Add a new api `reduce_scatter` and `all_to_all_single` since we need it in the rowwise sharding.
2. Modify the `all_to_all` api to make sure it consistent with the ones in distributed_c10d.py.
3. Found the cpp input params of `reduce_scatter` is missing input param, added more unit test to cover these cases.
4. Sync the NN test from gloo to nccl.

Differential Revision: [D32569674](https://our.internmc.facebook.com/intern/diff/D32569674/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @ezyang @gchanan

Commit:
Auto grad linear backward enabling

To enable the auto grad for the sharded linear, we find we need to make some changes to the current nn function api (c10d api with auto grad enabled). So we made the following several changes:

1. Add a new api `reduce_scatter` since we need it in the rowwise sharding.
2. Modify the `all_to_all` api to make sure it consistent with the ones in distributed_c10d.py.
3. Found the cpp input params of `reduce_scatter` is missing input param, added more unit test to cover these cases.
4. Sync the NN test from gloo to nccl.

Differential Revision: [D32569674](https://our.internmc.facebook.com/intern/diff/D32569674/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Distributed] Enable Reduce Scatter and modify all_to_all for sharded linear with more test cases."


To enable the auto grad for the sharded linear, we find we need to make some changes to the current nn function api (c10d api with auto grad enabled). So we made the following several changes:

1. Add a new api `reduce_scatter` since we need it in the rowwise sharding.
2. Modify the `all_to_all` api to make sure it consistent with the ones in distributed_c10d.py.
3. Found the cpp input params of `reduce_scatter` is missing input param, added more unit test to cover these cases.
4. Sync the NN test from gloo to nccl.

Differential Revision: [D32569674](https://our.internmc.facebook.com/intern/diff/D32569674/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -79,6 +79,28 @@ def reduce(tensor, dst, op=dist.ReduceOp.SUM, group=dist.group.WORLD):
     return _Reduce.apply(dst, op, group, tensor)
 
 
+def reduce_scatter(output, input_tensors, op=dist.ReduceOp.SUM, group=dist.group.WORLD):
```
Lets have 1:1 parity with this API: https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L2408 (input_list is the parameter name and not input_tensors)

fduwjj replied:
sure.

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -94,20 +116,23 @@ def all_gather(tensor, group=dist.group.WORLD):
 
 
-def all_to_all(tensors, group=dist.group.WORLD):
+def all_to_all(output_tensor_list, input_tensor_list, group=dist.group.WORLD):
```
I'm wondering if we should support `all_to_all_single` as well here? The reason is that I believe this `all_to_all` API might probably lead to some additional copies in our sharded linear implementation.

fduwjj replied:
Sure. I think this shouldn't be hard. I can do that.

pritamdamania87 replied:
@mrshenli Was wondering if you had some thoughts around these BC breaking API changes? It makes these APIs more inline with the ones in distributed_c10d.py and I was wondering if this is fine since my understanding is that we haven't really officially released these APIs? (despite having them on our docs page).

mrshenli replied:
Yep, they are not yet officially release, and it should be fine to introduce BC-breaking changes. I created a PR to remove those APIs from our docs. 

Also cc @cbalioglu

mrshenli replied:
oops, forgot to add a pointer to the PR: https://github.com/pytorch/pytorch/pull/69011

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -207,6 +232,20 @@ def backward(ctx, grad_output):
+    @staticmethod
+    def backward(ctx, grad_output):
+        result = _AllGather.apply(ctx.group, grad_output.contiguous())
+        print("backward ", grad_output, result)
```
nit: remove this?

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -207,6 +232,20 @@ def backward(ctx, grad_output):
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        result = _AllGather.apply(ctx.group, grad_output.contiguous())
```
Do we need to make this contiguous?

fduwjj replied:
Unfortunately, yes. So in the unit test which I write in this PR, the test passes. But for sharded linear test cases, it throws this exception:

> RuntimeError: Tensors must be non-overlapping and dense.

So I have to add this. And I don't know whether I am missing anything here.

pritamdamania87 replied:
Ah I see, do we know where this non-contiguous tensor comes from in sharded linear? We have mostly focussed on enabling functionality currently, but when we need to make the sharded linear implementation performant, we will have to look into things like these to avoid additional contiguous calls.

Also, if this is coming from sharded linear I'm wondering if it can be fixed there and in the core collective calls we don't have any `.contiguous()` calls?

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -207,6 +232,20 @@ def backward(ctx, grad_output):
+    def backward(ctx, grad_output):
+        result = _AllGather.apply(ctx.group, grad_output.contiguous())
+        print("backward ", grad_output, result)
+        return (None, None, None) + _AllGather.apply(ctx.group, grad_output.contiguous())
```
Shouldn't we return `result` here? Why are we calling allgather again?

fduwjj replied:
opps... Sorry I forgot removing the debugging code...

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -219,18 +258,26 @@ def forward(ctx, group, tensor):
         ctx.group = group
-        out_tensor_list = [
-            torch.empty_like(tensors[i]) for i in range(dist.get_world_size(group=group))
+        ctx.input_tensor_list = [
+            torch.empty_like(tensors[i])
+            for i in range(dist.get_world_size(group=group))
```
Since this is only used during backward, why not just save the desired sizes in the ctx and then allocate memory only in the backward method?

fduwjj replied:
I see, that makes sense.

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -94,20 +116,23 @@ def all_gather(tensor, group=dist.group.WORLD):
     Arguments:
         tensors (list[Tensor]): List of tensors to scatter one per rank.
         group (ProcessGroup, optional): The process group to work on.
+        out_tensor_list (list[Tensor]): list of tensors to gather one per rank.
```
We should document the behavior when `out_tensor_list` is None.

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -219,18 +258,26 @@ def forward(ctx, group, tensor):
+            torch.empty_like(tensors[i])
+            for i in range(dist.get_world_size(group=group))
         ]
+        if out_tensor_list is None:
```
I'm assuming we do this for BC reasons since the distributed_c10d.py all_to_all doesn't support this as being `None`?

fduwjj replied:
Yes, maybe we should make this consistent with `distributed_c10d.py`?

pritamdamania87 replied:
Based on @mrshenli's comment I think it is fine to make BC breaking changes here. So let's try to make everything consistent with `distributed_c10d.py`.

pritamdamania87 commented on test/distributed/test_c10d_spawn_nccl.py:
```diff
@@ -64,8 +83,170 @@ def test_shared_allgather_nccl(self):
+        @sandcastle_skip_if(
+            not _torch_dist_nn_available, "torch.distributed.nn is not available"
+        )
+        def test_broadcast(self):
+            store = c10d.FileStore(self.file_name, self.world_size)
+            # This is required because these functions calls directly to the .dist and needs
+            # the world to be initialized
+            c10d.init_process_group(
+                store=store, rank=self.rank, world_size=self.world_size, backend="nccl"
+            )
+            device = torch.device(f"cuda:{self.rank}")
+            x = torch.ones(5, 5, device=device) + self.rank
+            x.requires_grad = True
+            y = torch.distributed.nn.broadcast(x, 1)
+            self.assertEqual(y, 1 + torch.ones(5, 5))
+            z = y.sin().sum()
+            z.backward()
+            # We can't check the gradient of communications numerically so we have to do some calculations
+            if self.rank == 1:
+                self.assertEqual(x.grad, 2 * torch.cos(x))
+            elif self.rank == 0:
+                self.assertEqual(x.grad, torch.zeros(5, 5, device=device))
```
Can we reuse this code between `test_c10d_spawn_gloo` and `test_c10d_spawn_nccl` instead of repeating it here? There is `test_c10d_common.py` where we can probably put the common logic.

fduwjj replied:
Sure, let me refactor it a little bit and move the commonly supported module to the common.

Commit:
Update on "[PyTorch][Distributed] Enable Reduce Scatter and modify all_to_all for sharded linear with more test cases."


To enable the auto grad for the sharded linear, we find we need to make some changes to the current nn function api (c10d api with auto grad enabled). So we made the following several changes:

1. Add a new api `reduce_scatter` since we need it in the rowwise sharding.
2. Modify the `all_to_all` api to make sure it consistent with the ones in distributed_c10d.py.
3. Found the cpp input params of `reduce_scatter` is missing input param, added more unit test to cover these cases.
4. Sync the NN test from gloo to nccl.

Differential Revision: [D32569674](https://our.internmc.facebook.com/intern/diff/D32569674/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

fduwjj commented:
Addressed the feedback and added the `all_to_all_single`. Move the test of common ops to `test_c10d_spawn.py`.

Commit:
Update on "[PyTorch][Distributed] Enable Reduce Scatter and modify all_to_all for sharded linear with more test cases."


To enable the auto grad for the sharded linear, we find we need to make some changes to the current nn function api (c10d api with auto grad enabled). So we made the following several changes:

1. Add a new api `reduce_scatter` and `all_to_all_single` since we need it in the rowwise sharding.
2. Modify the `all_to_all` api to make sure it consistent with the ones in distributed_c10d.py.
3. Found the cpp input params of `reduce_scatter` is missing input param, added more unit test to cover these cases.
4. Sync the NN test from gloo to nccl.

Differential Revision: [D32569674](https://our.internmc.facebook.com/intern/diff/D32569674/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

mrshenli commented on torch/distributed/nn/functional.py:
```diff
@@ -88,26 +110,60 @@ def all_gather(tensor, group=dist.group.WORLD):
     return gathered list of tensors in output list.
 
     Arguments:
         tensors (list[Tensor]): List of tensors to scatter one per rank.
         group (ProcessGroup, optional): The process group to work on.
+        out_tensor_list (list[Tensor]): list of tensors to gather one per rank.
```
The args in the docstring do not seem to match the ones in the code?

docstring: tensors, group, out_tensor_list
code: output_tensor_list, input_tensor_list, group

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -219,19 +289,22 @@ def forward(ctx, group, tensor):
+            tensors[i].size() for i in range(dist.get_world_size(group=group))
         ]
-        reqs = [None] * dist.get_world_size(group=group)
+        out_tensor_list = [tensor.contiguous() for tensor in out_tensor_list]
```
I'm wondering if it is possible to avoid have `.contiguous` calls at all in the nn/functional.py APIs and deal with this in the sharded linear implementation itself?

For a general autograd enabled collective API it is better to throw an exception regarding contiguous tensors are not supported instead of adding overhead of the `contiguous` conversion.

Although, this might be tricky since the forward pass can be easily controlled in the sharded linear implementation but the backward pass is harder.

fduwjj replied:
Yeah, the test failure happens in the backward stage which might be out of our own control.

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -241,12 +314,49 @@ def forward(ctx, group, *tensors):
-            dist.all_to_all(out_tensor_list, list(tensors), group=group)
+            dist.all_to_all(
+                out_tensor_list,
+                list(tensor.contiguous() for tensor in tensors),
```
We have contiguous here as well, similar comments as above apply here too.

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -241,12 +314,49 @@ def forward(ctx, group, *tensors):
+    def forward(ctx, group, output, output_split_sizes, input_split_sizes, input):
+        ctx.group = group
+        ctx.input_size = input.size()
+        ctx.output_split_sizes_grad = input_split_sizes
+        ctx.input_split_sizes_grad = output_split_sizes
```
nit: Why are these called `grad`? There is no grad here? Maybe we just call them `ctx.output_split_sizes`?

fduwjj replied:
sure.

Commit:
Update on "[PyTorch][Distributed] Enable Reduce Scatter, all_to_all_single and modify all_to_all for sharded linear with more test cases."


To enable the auto grad for the sharded linear, we find we need to make some changes to the current nn function api (c10d api with auto grad enabled). So we made the following several changes:

1. Add a new api `reduce_scatter` and `all_to_all_single` since we need it in the rowwise sharding.
2. Modify the `all_to_all` api to make sure it consistent with the ones in distributed_c10d.py.
3. Found the cpp input params of `reduce_scatter` is missing input param, added more unit test to cover these cases.
4. Sync the NN test from gloo to nccl.

Differential Revision: [D32569674](https://our.internmc.facebook.com/intern/diff/D32569674/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

fduwjj commented on torch/distributed/nn/functional.py:
```diff
@@ -79,6 +79,28 @@ def reduce(tensor, dst, op=dist.ReduceOp.SUM, group=dist.group.WORLD):
+def reduce_scatter(output, input_list, op=dist.ReduceOp.SUM, group=dist.group.WORLD):
+    """
+    Reduces the tensor data across all machines, then scatters a list of
+        tensors to all processes in a group.
```
I see. I thought the indentation here is needed.

fduwjj commented on torch/distributed/nn/functional.py:
```diff
@@ -79,6 +79,28 @@ def reduce(tensor, dst, op=dist.ReduceOp.SUM, group=dist.group.WORLD):
 
+def reduce_scatter(output, input_list, op=dist.ReduceOp.SUM, group=dist.group.WORLD):
+    """
+    Reduces the tensor data across all machines, then scatters a list of
+        tensors to all processes in a group.
+
+    Each process is going to receive its corresponding final result.
+
+    Arguments:
+        output (Tensor): Output of the collective.
+        input_list (list[Tensor]): List of tensors to be aggregated and scatter one per rank.
+        op (optional): One of the values from
+            ``torch.distributed.ReduceOp``
+            enum.  Specifies an operation used for element-wise reductions.
+        group (ProcessGroup, optional): The process group to work on.
+
+    Returns:
+        Tensor: Output of the collective.
```
Sure.

Commit:
Update on "[PyTorch][Distributed] Enable Reduce Scatter, all_to_all_single and modify all_to_all for sharded linear with more test cases."


To enable the auto grad for the sharded linear, we find we need to make some changes to the current nn function api (c10d api with auto grad enabled). So we made the following several changes:

1. Add a new api `reduce_scatter` and `all_to_all_single` since we need it in the rowwise sharding.
2. Modify the `all_to_all` api to make sure it consistent with the ones in distributed_c10d.py.
3. Found the cpp input params of `reduce_scatter` is missing input param, added more unit test to cover these cases.
4. Sync the NN test from gloo to nccl.

Differential Revision: [D32569674](https://our.internmc.facebook.com/intern/diff/D32569674/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

fduwjj commented:
Address the comment from reviewers and keep the contiguous in the backward pass because ops like `SumBackward0` and  `TransposeBackward0 ` returns non-contiguous results.

pritamdamania87 commented on torch/distributed/nn/functional.py:
```diff
@@ -316,7 +308,7 @@ def forward(ctx, group, out_tensor_list, *tensors):
             dist.all_to_all(
                 out_tensor_list,
-                list(tensor.contiguous() for tensor in tensors),
+                list(tensors),
```
We don't need `.contiguous` here?

fduwjj replied:
No, this is forward and we don't need it. As long as we add contiguous in the backward, we are good to go.

Commit:
Update on "[PyTorch][Distributed] Enable Reduce Scatter, all_to_all_single and modify all_to_all for sharded linear with more test cases."


To enable the auto grad for the sharded linear, we find we need to make some changes to the current nn function api (c10d api with auto grad enabled). So we made the following several changes:

1. Add a new api `reduce_scatter` and `all_to_all_single` since we need it in the rowwise sharding.
2. Modify the `all_to_all` api to make sure it consistent with the ones in distributed_c10d.py.
3. Found the cpp input params of `reduce_scatter` is missing input param, added more unit test to cover these cases.
4. Sync the NN test from gloo to nccl.

Differential Revision: [D32569674](https://our.internmc.facebook.com/intern/diff/D32569674/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang ezyang gchanan

[ghstack-poisoned]

fduwjj commented:
Fix mypy error and rebase error of the backward function of `all_to_all_single`.

Commit:
Update on "[PyTorch][Distributed] Enable Reduce Scatter, all_to_all_single and modify all_to_all for sharded linear with more test cases."


To enable the auto grad for the sharded linear, we find we need to make some changes to the current nn function api (c10d api with auto grad enabled). So we made the following several changes:

1. Add a new api `reduce_scatter` and `all_to_all_single` since we need it in the rowwise sharding.
2. Modify the `all_to_all` api to make sure it consistent with the ones in distributed_c10d.py.
3. Found the cpp input params of `reduce_scatter` is missing input param, added more unit test to cover these cases.
4. Sync the NN test from gloo to nccl.

Differential Revision: [D32569674](https://our.internmc.facebook.com/intern/diff/D32569674/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang ezyang gchanan

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Distributed] Enable Reduce Scatter, all_to_all_single and modify all_to_all for sharded linear with more test cases."


To enable the auto grad for the sharded linear, we find we need to make some changes to the current nn function api (c10d api with auto grad enabled). So we made the following several changes:

1. Add a new api `reduce_scatter` and `all_to_all_single` since we need it in the rowwise sharding.
2. Modify the `all_to_all` api to make sure it consistent with the ones in distributed_c10d.py.
3. Found the cpp input params of `reduce_scatter` is missing input param, added more unit test to cover these cases.
4. Sync the NN test from gloo to nccl.

Differential Revision: [D32569674](https://our.internmc.facebook.com/intern/diff/D32569674/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang ezyang gchanan

[ghstack-poisoned]
